osu.Game/Online/SignalRUnionWorkaroundResolver.cs

```diff
@@ -0,0 +1,61 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Collections.Generic;
+using MessagePack;
+using MessagePack.Formatters;
+using MessagePack.Resolvers;
+using osu.Game.Online.Multiplayer;
+using osu.Game.Online.Multiplayer.MatchTypes.TeamVersus;
+
+namespace osu.Game.Online
+{
+    /// <summary>
+    /// Handles SignalR being unable to comprehend [Union] types correctly by redirecting to a known base (union) type.
+    /// See https://github.com/dotnet/aspnetcore/issues/7298.
+    /// </summary>
+    public class SignalRUnionWorkaroundResolver : IFormatterResolver
+    {
+        public static readonly MessagePackSerializerOptions OPTIONS =
+            MessagePackSerializerOptions.Standard.WithResolver(new SignalRUnionWorkaroundResolver());
+
+        private static readonly Dictionary<Type, IMessagePackFormatter> formatter_map = new Dictionary<Type, IMessagePackFormatter>
+        {
+            { typeof(TeamVersusUserState), new TypeRedirectingFormatter<TeamVersusUserState, MatchUserState>() },
+            { typeof(TeamVersusRoomState), new TypeRedirectingFormatter<TeamVersusRoomState, MatchRoomState>() },
+            { typeof(ChangeTeamRequest), new TypeRedirectingFormatter<ChangeTeamRequest, MatchUserRequest>() },
+
+            // These should not be required. The fallback should work. But something is weird with the way caching is done.
+            // For future adventurers, I would not advise looking into this further. It's likely not worth the effort.
+            { typeof(MatchUserState), new TypeRedirectingFormatter<MatchUserState, MatchUserState>() },
+            { typeof(MatchRoomState), new TypeRedirectingFormatter<MatchRoomState, MatchRoomState>() },
+            { typeof(MatchUserRequest), new TypeRedirectingFormatter<MatchUserRequest, MatchUserRequest>() },
+            { typeof(MatchServerEvent), new TypeRedirectingFormatter<MatchServerEvent, MatchServerEvent>() },
+        };
+
+        public IMessagePackFormatter<T> GetFormatter<T>()
+        {
+            if (formatter_map.TryGetValue(typeof(T), out var formatter))
+                return (IMessagePackFormatter<T>)formatter;
+
+            return StandardResolver.Instance.GetFormatter<T>();
+        }
+
+        public class TypeRedirectingFormatter<TActual, TBase> : IMessagePackFormatter<TActual>
```

Comment:
You could drop three out of four casts below by using type constraints here:

```suggestion
        public class TypeRedirectingFormatter<TActual, TBase> : IMessagePackFormatter<TActual>
            where TActual : class, TBase
            where TBase : class
```

leaving only the downcast of `TBase` to `TActual` on deserialisation.

Although I suppose that both locks in this formatter to only be usable for the inheritance issues, and makes it less explicit (you have to notice that this only works with the casts dropped because `baseFormatter` is of type `IMessagePackFormatter<TBase>`).